platform-sdk/docs/proposals/state-lifecycle/proposal.md

```diff
@@ -0,0 +1,268 @@
+   - Handling consensus rounds  
+   - Sealing consensus rounds  
+   - Keeping the `freeze time` parameter up to date  
+   - Providing the state for signing  
+   - Initializing the state by extracting it from the signed state  
+
+## Proposed Changes
+
+All lifecycle-related methods from `com.swirlds.state.State` should be moved to a separate class responsible for managing the state lifecycle. This will provide a clearer separation of concerns.  
+
+- `com.swirlds.state.State` should focus solely on accessing and modifying the state.  
+- The state lifecycle will be managed externally.
+- Its hash will be computed asynchronously and managed externally.
+
+```java
+public interface State {
+   @NonNull
+   ReadableStates getReadableStates(@NonNull String serviceName);
+
+   @NonNull
+   WritableStates getWritableStates(@NonNull String serviceName);
+
+   default void registerCommitListener(@NonNull final StateChangeListener listener)l;
+   
+   @Nullable
+   Future<Hash> getHashFuture();
+}
+```
+#### Notes
+- The signature of the `State#getHash()` method will change to `Future<Hash> getHash()` to accommodate asynchronous hash computation.
+- The `State` interface will no longer include the `copy`, `computeHash`, `createSnapshot`, or `loadSnapshot` methods.
+
+The `SwirldStateManager` class requires refactoring to separate its two distinct responsibilities. `swirlds-state-impl` should not include any platform-specific details but should provide the necessary mechanisms for the platform to interact with the state and its lifecycle.
+
+### `StateLifecycleManager` Interface
+
+To achieve this, `swirlds-state-api` will define a `StateLifecycleManager` interface, with an implementation `StateLifecycleManagerImpl` in `swirlds-state-impl`. The interface will handle:
+
+- Creating/loading an initial state. If no snapshot is available, a new state will be created.  
+- Managing references to the latest immutable state and the latest mutable state.
+- Resetting the mutable state to the latest immutable state.
+- Restricting mutability to a single state object. Specifically, only one state can be mutable at a time.  
+- Loading snapshots from disk. States loaded from snapshots are immutable.  
+
+The interface will look something like this:
+
+```java
+public interface StateLifecycleManager {
+    /**
+     * Get the mutable state.
+     */
+    State getMutableState();
+
+   /**
+    * Reset the mutable state to the latest immutable state.
+    */
+   void resetMutableState();
+
+   /**
+    * Creates a snapshot for the latest immutable state. 
+    * 
+    * @param targetPath The path to save the snapshot.
+    */
+    State createSnapshot(final @NonNull Path targetPath);
+
+    /**
+     * Creates a mutable copy of the state. The previous mutable state becomes immutable,
+     * replacing the latest immutable state. 
+     * 
+     * @return a mutable copy of the state
+     */
+    State copyMutableState();
```

Comment:
I think this should to return the immutable state that was used as the source of the copy operation. This is currently what the copy method in the `SwirldsStateManager` class returns because that is what is used for the signed state. It is the thing that moves forward in the lifecycle. If it is not returned by this method, the caller code will look something like this:

```
State soonToBeImmutableState = sslm.getMutableState();
sslm.copyMutableState();
SignedState ss = new SignedState(soonToBeImmutableState); // now is immutable
```

If it returns the immutable state, then it looks like this:

```
State completedState = sslm.copyMutableState();
SignedState ss = new SignedState(completedState);
```

Reply:
Yes, that's a good point. I'll make this change.